### ISSUE TITLE
VTU with cells of different types

### ISSUE THREAD

**Issue by FabienSalmon** [ornladios/ADIOS2#3045]
Hello,

I use adios2 to make, for instance, a 2D square with rectangle cells and it works perfectly when I open it with paraview.

To do that I use a type equal to 8 and I use the following lines to add this variable (in fortran):

```
types = 8
call adios2_define_variable(varArray, io, "types", adios2_type_integer4, ierr)
call adios2_put(engine, varArray, types, ierr)
```

bpls -lav gives the following output:

```
File info:
  of variables:  7
  of attributes: 1
  statistics:    Min / Max 

  int32_t  NumOfElements     {2} = 5130 / 7866
  int32_t  NumOfVertices     {2} = 5290 / 8050
  int64_t  connectivity      [2]*{__, 5} = 0 / 8049
  int32_t  dimension         attr   = 2
  float    ib_distance_serp  [2]*{__} = -0.00100592 / 0.00333333
  float    temperature       [2]*{__} = 0 / 0
  int32_t  types             scalar = 8
  double   vertices          [2]*{__, 2} = 0.1 / 0.48
```

Now, in order to use other kinds of cells, I tried to pass the types as an array. Thus, I created an array of the same size as the connectivity (so the number of cells) and did:

```
allocate(types(size(connectivity,2)))
types = 8
count_dims5 = [ size(connectivity,2) ]
call adios2_define_variable(varArray, io, "types", adios2_type_integer4, 1, adios2_null_dims, adios2_null_dims, count_dims5,  adios2_constant_dims, ierr)
call adios2_put(engine, varArray, types, ierr)
```

bpls -lav gives the following output:

```
File info:
  of variables:  7
  of attributes: 1
  statistics:    Min / Max 

  int32_t  NumOfElements     {2} = 5130 / 7866
  int32_t  NumOfVertices     {2} = 5290 / 8050
  int64_t  connectivity      [2]*{__, 5} = 0 / 8049
  int32_t  dimension         attr   = 2
  float    ib_distance_serp  [2]*{__} = -0.00100592 / 0.00333333
  float    temperature       [2]*{__} = 0 / 0
  int32_t  types             [2]*{__} = 8 / 8
  double   vertices          [2]*{__, 2} = 0.1 / 0.48
```

I expected this to work exactly in the same way as previously, but it does not. I can show the points of the geometry, but not the cells, paraview does not want even if it does not crash. Did I miss something? In other words, do we need to add something else to use an array for the types instead of a scalar? Or is this a bug in adios2?

Thank you for your replies

**Comment by williamfgc** [comment]
@FabienSalmon thanks for bringing this up. Are you using the VTX reader or the core reader in ParaView?  Only VTX supports unstructured data and it uses a single cell type defined by a `uint32_t`. I'll follow the example in the [docs](https://adios2.readthedocs.io/en/latest/ecosystem/visualization.html). Let me know if you have any questions. Hope it helps.

**Comment by FabienSalmon** [comment]
Thank you for your quick reply.
Indeed, I use the VTX reader and read the bp4 directory in paraview.

So, if I understand well, it is not possible to use different kinds of cells so far? So, even if I have some square cells, I have to use polygonal cells in 2D if I have at least one polygon among my cells (and define the type as a scalar)?

**Comment by williamfgc** [comment]
Hi @FabienSalmon that's correct. Does VTK support that use-case? It's been a while, I have to check their file format docs where the schemas are defined. Thanks.

**Comment by FabienSalmon** [comment]
Ok, thank you for the reply.
I do not think VTK supports this case but not sure.

**Comment by FabienSalmon** [comment]
I relaunch the discussion because after some tries, it does not seem possible to deal with polygons with adios2. Actually, I think that having different sizes is a major problem for adios2. Besides, the given example on the website is based on mfem which do not encompass polygons or polyhedron.

So it seems even less possible to have a polyhedron mesh.

Am I correct ?

Thank you

**Comment by williamfgc** [comment]
@FabienSalmon do you have a plain text vtu file example (not adios2) that works for your case? That would be a good starting point since vtu is a rich environment and the adios2 VTX reader is just translating schemas to corresponding VTK data structures. Let me know.

**Comment by williamfgc** [comment]
@FabienSalmon yes, that's why a working example would be useful to see first if the VTK backend can actually support the case, that way we can accommodate the logic in the VTX VTK reader (that's where the functionality resides, not in adios2).

**Comment by FabienSalmon** [comment]
Ok, so to understand, you want a VTU file of a case?
If so, it will be far different from the structure of XML VTU but I can give you that

**Comment by FabienSalmon** [comment]
Here an XML file representing a simple mesh with 2 cells
[test2.zip](https://github.com/ornladios/ADIOS2/files/8098526/test2.zip)

Here a VTU file representing the same mesh
[test2vtu.zip](https://github.com/ornladios/ADIOS2/files/8098669/test2vtu.zip)

**Comment by FabienSalmon** [comment]
I think I might have an explanation but not sure, please, tell me what do you think.

Adios2 simply write the arrays. So, when we write the connectivity array if there are cells composing of a different number of points, we have for instance:

`4 1 2 3 4 0`
`3 2 4 5 0 0`
`5 1 4 5 6 2`


where the first column is the number of vertices. We must have some zeros (or other numbers) in the connectivity array. But, to be read by paraview, we might need to have

`4 1 2 3 4`
`3 2 4 5` 
`5 1 4 5 6 2`


without the zeros like all the examples provided on the web in VTU. But we cannot remove these zeros in the framework of adios2. Do you think the problem stems from that ?

Thank you

**Comment by williamfgc** [comment]
@FabienSalmon this scenario was never supported nor was in the roadmap. Let's treat it as a new feature since the visualization was written for specific cases. I'll try to find some time to explore options in the VTX reader. Thanks!

**Comment by FabienSalmon** [comment]
Ok thank you for the reply.

Hope you will find time to do that, I think it would be of great interest for adios2.
